### PR TITLE
Clarify And Unify doc 2 - Return of the docs

### PR DESCRIPTION
**Pull request by ChrisPenner**
See https://github.com/ramda/ramda/pull/1891,
Doing a run over examples clarifying things,

Of note in this batch is that the type signatures for some of the compose functions didn't always show that the first function could take multiple arguments. Quick question on that, is composeK supposed to allow multiple arguments for the first function? I couldn't find it mentioned anywhere.

Also I noticed that the multiple argument functionality of `compose` doesn't seem to be tested 😢 
As usual, I'll squash this down when it's ready to merge.

### CODE REVIEW

File: src/comparator.js [hunk 1]
```diff
@@ -10,15 +10,15 @@ var _curry1 = require('./internal/_curry1');
  * @since v0.1.0
  * @category Function
  * @sig (a, b -> Boolean) -> (a, b -> Number)
- * @param {Function} pred A predicate function of arity two.
- * @return {Function} A Function :: a -> b -> Int that returns `-1` if a < b, `1` if b < a, otherwise `0`.
+ * @param {Function} pred A predicate function of arity two. Return `true` if the first argument is less than the second, `false` otherwise
```
Comment (CrossEye) [inline src/comparator.js]: Perhaps joining these two into

    A predicate function of arity two which will return `true` if the first argument 
    is less than the second, `false` otherwise

File: src/chain.js [hunk 2]
```diff
@@ -17,9 +17,9 @@ var map = require('./map');
  * @since v0.3.0
  * @category List
  * @sig Chain m => (a -> m b) -> m a -> m b
- * @param {Function} fn
- * @param {Array} list
- * @return {Array}
+ * @param {Function} fn The function to map with
+ * @param {Array} list The list to map over
+ * @return {Array} An array formed by concatenating the results of mapping over `list` with `fn`.
```
Comment (buzzdecafe) [inline src/chain.js]: perhaps "The result of flat-mapping `list` with `fn`"
?

File: src/complement.js [hunk 3]
```diff
@@ -3,15 +3,10 @@ var not = require('./not');
...
+ * Takes a function `f` and returns a function `g` such that if called with the same arguments
+ * when `f` returns `true`, `g` returns `false` and vice versa
  *
- *   - applying `g` to zero or more arguments will give __true__ if applying
- *     the same arguments to `f` gives a logical __false__ value; and
- *
- *   - applying `g` to zero or more arguments will give __false__ if applying
- *     the same arguments to `f` gives a logical __true__ value.
- *
- * `R.complement` will work on all other functors as well.
```
Comment (CrossEye) [inline src/complement.js]: This is trickier.  The original is wordy, probably far too wordy, but it's also more accurate than the replacement.

The trouble is that this is not restricted to working with functions that return booleans.  It deals with Javascript's truthy and falsey ideas, so, for instance, `R.complement(R.identity)` would return `true` for all of these values: `false`, `null`, `undefined`, `''`, `NaN`, `+0`, and `-0`.

File: src/clamp.js [hunk 4]
```diff
@@ -10,14 +10,14 @@ var _curry3 = require('./internal/_curry3');
  * @since v0.20.0
  * @category Relation
  * @sig Ord a => a -> a -> a -> a
- * @param {Number} minimum number
- * @param {Number} maximum number
- * @param {Number} value to be clamped
- * @return {Number} Returns the clamped value
+ * @param {Number} minimum The lower range of the clamp
+ * @param {Number} maximum The upper range of the clamp
```
Comment (buzzdecafe) [inline src/clamp.js]: "limit" may be better than "range" here

we should probably indicate if these limits are inclusive or not. `range` for example is inclusive at the bottom, but not at the top.

File: src/comparator.js [hunk 5]
```diff
@@ -10,15 +10,15 @@ var _curry1 = require('./internal/_curry1');
...
+ * @return {Function} A Function :: a -> b -> Int that returns `-1` if a < b, `1` if b < a, otherwise `0`
  * @example
  *
- *      var cmp = R.comparator((a, b) => a.age < b.age);
+ *      var ageComparator = R.comparator((a, b) => a.age < b.age);
```
Comment (buzzdecafe) [inline src/comparator.js]: `byAge` would read well

File: src/comparator.js [hunk 6]
```diff
@@ -10,15 +10,15 @@ var _curry1 = require('./internal/_curry1');
...
  *      var people = [
  *        // ...
  *      ];
- *      R.sort(cmp, people);
+ *      var peopleByIncreasingAge = R.sort(ageComparator, people);
```
Comment (buzzdecafe) [inline src/comparator.js]: ... because here it would be `sort(byAge, ...)`

File: src/compose.js [hunk 7]
```diff
@@ -12,15 +12,18 @@ var reverse = require('./reverse');
  * @memberOf R
  * @since v0.1.0
  * @category Function
- * @sig ((y -> z), (x -> y), ..., (o -> p), ((a, b, ..., n) -> o)) -> ((a, b, ..., n) -> z)
- * @param {...Function} functions
+ * @sig ((x -> z), ..., (b -> c), ((a1, a2, ...) -> b), (a1, a2, ...)) -> ((a1, a2, ...) -> z)
```
Comment (buzzdecafe) [inline src/compose.js]: the original sig looks correct, i am having a hard time parsing this one

File: src/compose.js [hunk 8]
```diff
@@ -12,15 +12,18 @@ var reverse = require('./reverse');
...
+ * @param {...Function} ...functions The functions to compose
  * @return {Function}
  * @see R.pipe
  * @example
  *
- *      var f = R.compose(R.inc, R.negate, Math.pow);
+ *      var classyGreeting = (firstName, lastName) => "Hello, I'm " + lastName + ", " + firstName + " " + lastName
+ *      var yellGreeting = R.compose(R.toUpper, classyGreeting);
+ *      yellHello('Bob', 'Smith'); //=> "HELLO, I'M SMITH, BOB SMITH"
```
Comment (buzzdecafe) [inline src/compose.js]: 'James', 'Bond'

File: src/compose.js (hunk 7 above)
Comment (CrossEye) [inline src/compose.js]: I would like to simplify this sig, but I don't think that quite captures it correctly.  There are at least two things that confuse me.  I don't know why there is a switch from an initial `y` to an initial `x`.  And I don't understand what the second `(a1, a2, ...)` is about.

I'm also quite mixed about using the `a1` / `a2` notation.  I would love to be able to get rid of it altogether, as I think it confuses things.  But there seem to be times we need it.

To that end, I've been thinking lately about switching how we document these variadic signatures.  But the ones that are variadic in two ways are harder.  My thought was to express the ellipsis as applying only to a collection of full signatures.  For example,

```js
// compose :: (b -> c) -> (a -> b) -> (a -> c) 
//         :: (c -> d) -> (b -> c) -> (a -> b) -> (a -> d)
//         :: (d -> e) -> (c -> d) -> (b -> c) -> (a -> b) -> (a -> e)
//         :: ...
```

The trouble, of course, is that `compose` is variadic in a second dimension:

```js
// compose :: (b -> c) -> (a -> b) -> (a -> c) 
//         :: (c -> d) -> (a -> b -> c) -> (a -> b -> d)
//         :: (d -> e) -> (a -> b -> c -> d) -> (a -> b -> c -> e)
//         :: ...
```

And while we could certainly write out an example of combining these, there is no way to clearly express this in a linear pattern like those above.

```js
//         :: (f -> g) -> (e -> f) -> (d -> e) -> (a -> b -> c -> d) -> (a -> b -> c -> g)
```

But I'm still tempted, as these lists seem a lot more intuitive than the in-line ellipsis.  I'll open up a separate issue for this, as this is not the place to discuss the notion in detail, but it did seem _a propos_, given this suggested signature change.

**Comment by CrossEye** [comment]
Thank you.  These (and the earlier similar PR) are great!

File: src/compose.js (hunk 7 above)
Comment (ChrisPenner) [inline src/compose.js]: It does kind of come down to whether we want to follow formal mathematic notation or just make something that looks readable.

I found the original (a, b, ..., n) to be pretty arbitrary, and it implied to me that the input argument types were somehow related to the other function types. Personally I found it clearer to show that only the first function was variadic by using a1, a2, ...

The y-> x switch is a typo, I'll fix that.

Comment (CrossEye) [inline src/compose.js]: Note that we're not following any formal notation, just hacking together whatever seem to be appropriate extensions to Hindley-Milner type signature descriptions to capture Javascript's differences.  I wrote a [lengthy article][wi] on how we're using them in Ramda.

>  Personally I found it clearer to show that only the first function was variadic by using a1, a2, ...

I certainly understand the temptation, but what bothers me is that now we have to represent an arbitrary type, not just the single-lowercase-letter tokens but also any of those followed by a number.  Would `a3` be fine on its own, or only in conjunction with `a1` and `a2`?  It just seems to make the rules more complex.  I don't totally abhor them, but I'm hoping we can find a way to avoid them.

  [wi]: https://github.com/ramda/ramda/wiki/Type-Signatures

**Comment by ChrisPenner** [comment]
This one should be almost good to go, I reverted the `compose` type signature, One question would be whether `composeP`'s signature should reflect that the first call can take an arbitrary number of arguments, but if no-one's worried about it then this can go through. @buzzdecafe

File: src/complement.js [hunk 9]
```diff
@@ -3,15 +3,10 @@ var not = require('./not');
 
 
 /**
- * Takes a function `f` and returns a function `g` such that:
+ * Takes a function `f` and returns a function `g` such that if called with the same arguments
+ * when `f` returns a "truthy" value, `g` returns `false` and when `f` returns a "falsy" value `g` returns `true`.
```
Comment (CrossEye) [inline src/complement.js]: Nice!

**Comment by CrossEye** [comment]
LGTM.  Anyone else before we merge?

**Comment by buzzdecafe** [comment]
I was  hoping @davidchambers would weigh in; copy-editing is one of his many strengths.

**Comment by CrossEye** [comment]
I should really have just pinged him.  That's who I'm hoping to see.

File: src/complement.js (hunk 3 above)
Comment (davidchambers) [inline src/complement.js]: I prefer the current wording, but I think I wrote it so it's not surprising it reads well to me. ;)

File: src/compose.js [hunk 10]
```diff
@@ -13,14 +13,17 @@ var reverse = require('./reverse');
...
- * @param {...Function} functions
+ * @param {...Function} ...functions The functions to compose
  * @return {Function}
  * @see R.pipe
  * @example
  *
- *      var f = R.compose(R.inc, R.negate, Math.pow);
+ *      var classyGreeting = (firstName, lastName) => "The name's " + lastName + ", " + firstName + " " + lastName
+ *      var yellGreeting = R.compose(R.toUpper, classyGreeting);
+ *      yellHello('James', 'Bond'); //=> "THE NAME'S BOND, JAMES BOND"
```
Comment (davidchambers) [inline src/compose.js]: `s/Hello/Greeting/`

**Comment by ChrisPenner** [comment]
@davidchambers Good to go?

**Comment by buzzdecafe** [comment]
I think you corrected what he found, so I am going to go ahead and merge. And thanks
